### ISSUE TITLE
report an error when `@final` decorator is used on a private method

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ty#3140]
### `main.py`

```py
from typing import final, override


class Foo:
    @final # no error. should say something like "@final decorator is useless on private methods"
    def __foo(self):
        ...

class Bar(Foo):
    @override
    def __foo(self): # correct error: Method `__foo` is decorated with `@override` but does not override anything
        ...
```

**Comment by AlexWaygood** [comment]
It is _possible_ to override a private method, just "quite hard" to do so:

```pycon
>>> class Foo:
...     def __private(self):
...         print("__private called")
...         
>>> Foo()._Foo__private()
__private called
>>> class Bar(Foo):
...     def _Foo__private(self):
...         print("overridden")
...         
>>> Bar()._Foo__private()
overridden
```

Rather than implementing any special casing for private methods here and in other places, I think we should just implement Python's name mangling semantics more generally, which is https://github.com/astral-sh/ty/issues/645.

I don't particularly see why we should go out of our way to emit a diagnostic for `@final` being applied to a name-mangled method. It's an odd thing to do (you're not _meant_ to override these methods generally, so it's sort-of unnecessary), but it's not _wrong_.

**Comment by charliermarsh** [comment]
I agree with @AlexWaygood.

**Comment by DetachHead** [comment]
> so it's sort-of unnecessary), but it's not *wrong.*

This is why I think there should be a warning about it. Since it's unnecessary, it's very likely to be a mistake

Would it make more sense as a ruff rule instead?

**Comment by AlexWaygood** [comment]
This might make sense as a lint rule, yes. It seems low-priority even there, though, to me. 

What motivated this issue report? Did you see a real-world example of somebody doing this?

**Comment by DetachHead** [comment]
During a refactor I accidentally changed a protected method to a private method and forgot to delete its `@final` decorator. I did notice before I committed it, but I believe that mistakes like this should be detected by a linter. Even if it doesn't cause any problems at runtime, it's still an obvious mistake that just makes your code more confusing to a reader if it isnt caught.

**Comment by DetachHead** [comment]
> Rather than implementing any special casing for private methods here and in other places, I think we should just implement Python's name mangling semantics more generally, which is https://github.com/astral-sh/ty/issues/645.

The way I see it, name mangling is an implementation detail of private methods, not an actual feature that anyone should be relying on. As far as the type checker is concermed, the method is private and shouldn't be accessible from any subclass
